test_qmk_compiler.py

```diff
@@ -0,0 +1,337 @@
+        'key_count': 12,
+        'layout': [
+            {'label': 'k00', 'w': 1, 'x': 0, 'y': 0},
+            {'label': 'k01', 'w': 1, 'x': 1, 'y': 0},
+            {'label': 'k02', 'w': 1, 'x': 2, 'y': 0},
+            {'label': 'k10', 'w': 1, 'x': 0, 'y': 1},
+            {'label': 'k12', 'w': 1, 'x': 1, 'y': 1},
+            {'label': 'k20', 'w': 1, 'x': 0, 'y': 2},
+            {'label': 'k21', 'w': 1, 'x': 1, 'y': 2},
+            {'label': 'k22', 'w': 1, 'x': 2, 'y': 2},
+            {'label': 'k11', 'w': 1, 'x': 0, 'y': 3},
+            {'label': 'k30', 'w': 1, 'x': 0, 'y': 4},
+            {'label': 'k31', 'w': 1, 'x': 1, 'y': 4},
+            {'label': 'k32', 'w': 1, 'x': 2, 'y': 4}
+        ]
+    }
+
+
+### FIXME: Need to write a test for update_kb_redis.parse_config_h()
+
+
+def test_0019_parse_config_h_file_cluecard():
+    """Make sure the config.h parsing works.
+    """
+    config_h = update_kb_redis.parse_config_h_file('qmk_firmware/keyboards/clueboard/card/config.h')
+    assert config_h == {
+        'CONFIG_H': True,
+        'VENDOR_ID': '0xC1ED',
+        'PRODUCT_ID': '0x2330',
+        'DEVICE_VER': '0x0001',
+        'MANUFACTURER': 'Clueboard',
+        'PRODUCT': 'ATMEGA32U4 Firmware Dev Kit',
+        'DESCRIPTION': 'A small board to help you hack on QMK.',
+        'MATRIX_ROWS': '4',
+        'MATRIX_COLS': '3',
+        'MATRIX_ROW_PINS': '{ F0, F5, F4, B4 }',
+        'MATRIX_COL_PINS': '{ F1, F7, F6 }',
+        'UNUSED_PINS': True,
+        'DIODE_DIRECTION': 'ROW2COL',
+        'DEBOUNCING_DELAY': '20',
+        'BACKLIGHT_LEVELS': '6',
+        'IS_COMMAND()': '( \\',
+        'RGB_DI_PIN': 'E6',
+        'RGBLED_NUM': '4 // Number of LEDs',
+        'RGBLIGHT_HUE_STEP': '10',
+        'RGBLIGHT_SAT_STEP': '17',
+        'RGBLIGHT_VAL_STEP': '17'
+    }
+
+
+## FIXME: Write a test for update_kb_redis.parse_rules_mk
+
+
+def test_0021_parse_rules_mk_file_cluecard():
+    """Make sure the rules.mk parsing works.
+    """
+    rules_mk = update_kb_redis.parse_rules_mk_file('qmk_firmware/keyboards/clueboard/card/rules.mk')
+    assert rules_mk == {
+        'ARCH': 'AVR8',
+        'AUDIO_ENABLE': 'yes',
+        'BACKLIGHT_ENABLE': 'yes',
+        'BLUETOOTH_ENABLE': 'no',
+        'BOOTMAGIC_ENABLE': 'no',
+        'COMMAND_ENABLE': 'yes',
+        'CONSOLE_ENABLE': 'yes',
+        'EXTRAKEY_ENABLE': 'yes',
+        'F_CPU': '16000000',
+        'F_USB': '$(F_CPU)',
+        'MCU': 'atmega32u4',
+        'MIDI_ENABLE': 'no',
+        'MOUSEKEY_ENABLE': 'yes',
+        'NKRO_ENABLE': 'no',
+        'OPT_DEFS': '-DINTERRUPT_CONTROL_ENDPOINT -DBOOTLOADER_SIZE=4096',
+        'RGBLIGHT_ENABLE': 'yes',
+        'SLEEP_LED_ENABLE': 'no',
+        'UNICODE_ENABLE': 'no'
+    }
+
+
+def test_0022_default_key():
+    """Test update_kb_redis.default_key().
+    """
+    new_key1 = update_kb_redis.default_key()
+    new_key2 = update_kb_redis.default_key()
+    assert new_key1 == {'x': 3, 'y': 4, 'w': 1}
```

Comment:
what is a default key?

Reply:
It's basically an implementation detail. The state machine that translates a keymap.c into an info.json template uses that function to get the next key for the current row.